csrc/device_lower/pass/insert_syncs.cpp

```diff
@@ -1061,6 +1094,67 @@ class WarAsyncWaitInserter : private kir::ExprMutator {
     return pending_ops;
   }
 
+  // The for-loop for the k dimension in matmul expressions is circular
+  // buffered. TMA loads are launched to load mma operands into shared memory,
+  // which are consumed by mma expression. A WAR sync is necessary for the wgmma
+  // expession to avoid overwriting the circular buffer stage while it is being
+  // read.
+  //
+  // Special logic is required for warp specialized circular buffering because
+  // the TMA loads and wgmma ops are separated by an IfThenElse.
+  // kir::ExprMutator traverses the fusion in depth-wise order, so TMA loads in
+  // the LoadWarp are detected before the wgmma expressions in the ComputeWarp.
+  //
+  // This function inserts wgmma.commit_group and wgmma.wait_group expressions
+  // before the mbarrier::arrive, which allows load warp to launch next TMA
+  // load. First, we commit all the wgmma expressions issued in this iteration
+  // of the for-loop. Then, we wait for some number of wgmma expressions based
+  // on number of circular buffer stages and number of prefetch stages.
+  void handleComputeWarp(ForLoop* for_loop) {
+    NVF_ERROR(
+        for_loop->circularBufferLoopStage() ==
+            CircularBufferLoopStage::ComputeWarp,
+        "for_loop is not circular buffer compute warp");
+    NVF_ERROR(
+        !warp_specialized_async_inputs_in_current_scope_.empty(),
+        "No TMA loads were detected in LoadWarp.");
+    NVF_ERROR(
+        !warp_specialized_async_exprs_to_protect_.empty(),
+        "No WgMma operations were detected in ComputeWarp.");
```

Comment:
Does this change makes it impossible to use warp specialization on non-matmul kernel?

Reply:
You're right. I added a short-circuit path if there aren't any wgmma ops to protect.